File: 2dArray/1640.CheckArrayFormationThroughConcatenation/MatchDict.py
#!/usr/bin/env python3
def canFormArray(arr, pieces):
        for i in range(len(pieces)):
            if pieces[i][0] in arr:
                pos = arr.index(pieces[i][0])
                print(pos)
                for j in range(len(pieces[i])):
                    if pos >= len(arr) or pieces[i][j] != arr[pos]:
                        return False
                    pos = pos + 1
            else:
                return False
        return True

print(canFormArray([49, 18, 16], [[16, 18, 49]]))
print(canFormArray([2,4,5,6], [[4,5],[2],[6]]))

File: 2dArray/1640.CheckArrayFormationThroughConcatenation/MatchDict.js
/**
 * @param {number[]} arr
 * @param {number[][]} pieces
 * @return {boolean}
 */
/*
Matching-Dict Problem
1step: traverse pieces, check if arr include pieces[i][0] first element in inner pieces
2step: if not, return false; if yes, find position in arr; and check this inner pieces, if not remaining elements return false
*/
var canFormArray = function (arr, pieces) {
  for (let i = 0; i < pieces.length; i++) {
    if (arr.includes(pieces[i][0])) {
      let pos = arr.indexOf(pieces[i][0]);
      for (let j = 0; j < pieces[i].length; j++) {
        if (pieces[i][j] != arr[pos]) {
          return false;
        }
        pos++;
      }
    } else {
      return false;
    }
  }
  return true;
};
